### PR TITLE
Add GCP package

### PR DESCRIPTION
**Pull request by marc-gr**
## What does this PR do?

Adds GCP package

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/master/CONTRIBUTING.md#tips-for-building-integrations) and this pull request is aligned with them.
- [x] I have verified that all datasets collect metrics or logs.

## Screenshots

image

image

image

image

### CODE REVIEW

**Comment by andrewkroh** [comment]
I like that you converted as much as possible to Ingest Node.

The Google pub/sub service has an emulator. Perhaps the emulator can be loaded with our json logs and then the e2e tests pull from the emulator's topics.

The reason the logfile input exists in these GCP data streams is to allow for testing. By including it in the UI I think we make the usage less clear. We might be able to remove it since there is an alternative means of testing via the emulator. However some different services (like s3+sqs) don't have a good emulation option so logfile is still needed IMO. It might be useful to have a way for an input to be available only for testing.

**Comment by marc-gr** [comment]
> I like that you converted as much as possible to Ingest Node.
> 
> The Google pub/sub service has an emulator. Perhaps the emulator can be loaded with our json logs and then the e2e tests pull from the emulator's topics.
> 
> The reason the logfile input exists in these GCP data streams is to allow for testing. By including it in the UI I think we make the usage less clear. We might be able to remove it since there is an alternative means of testing via the emulator. However some different services (like s3+sqs) don't have a good emulation option so logfile is still needed IMO. It might be useful to have a way for an input to be available only for testing.

I will give it a try, and I opened an issue https://github.com/elastic/package-spec/issues/97 for what you are mentioning.

**Comment by marc-gr** [comment]
We have to wait for the changes in https://github.com/elastic/beats/pull/23215 to be able to use the emulator here

**Comment by andrewkroh** [comment]
jenkins, run tests

Now that `alternative_host` has been merged to Filebeat let's see how this goes.

**Comment by andrewkroh** [comment]
@marc-gr This needs an update to change  `config.yml` to  `test-pubsum.config.yml`. ref: https://github.com/elastic/package-spec/blob/6435a2f6b3e0b92c11ffb4531a0ad3f575093d3f/versions/1/data_stream/_dev/test/system/spec.yml#L6

File: packages/gcp/data_stream/firewall/agent/stream/gcp-pubsub.yml.hbs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  - {{tag}}
+  {{/each}}
+{{#contains tags "forwarded"}}
+publisher_pipeline.disable_host: true
+{{/contains}}
+processors:
+  - decode_json_fields:
+      fields: ["message"]
+      target: "json"
+  - community_id:
```
Comment (andrewkroh) [inline packages/gcp/data_stream/firewall/agent/stream/gcp-pubsub.yml.hbs:25]: Can you please try out the Ingest Node community_id processor as a replacement.

**Comment by adriansr** [comment]
For the logfile input, you may want to set `enabled: false` in the manifest.yml, so it's not enabled by default in the configuration.

**Review by mtojek — COMMENTED** [review]
@marc-gr  I see that you pushed a lot of Go code here, not sure if this is the expected way to go. In general it would be better to depend on DSLs/manifests/configs.

If you need to create some resources on GCP, you can try to use the Terraform service deployer: https://github.com/elastic/elastic-package/blob/master/docs/howto/system_testing.md#terraform-service-deployer

Here is the appliance for AWS: https://github.com/elastic/integrations/tree/master/packages/aws/data_stream/ec2_metrics/_dev/deploy/tf

**Comment by marc-gr** [comment]
> @marc-gr I see that you pushed a lot of Go code here, not sure if this is the expected way to go. In general it would be better to depend on DSLs/manifests/configs.
> 
> If you need to create some resources on GCP, you can try to use the Terraform service deployer: https://github.com/elastic/elastic-package/blob/master/docs/howto/system_testing.md#terraform-service-deployer
> 
> Here is the appliance for AWS: https://github.com/elastic/integrations/tree/master/packages/aws/data_stream/ec2_metrics/_dev/deploy/tf

Finally we enhanced `akroh/stream` to support GCP pubsub as an output, so we moved the go code there and just using docker here 👍

**Review by mtojek — COMMENTED** [review]
> Finally we enhanced akroh/stream to support GCP pubsub as an output, so we moved the go code there and just using docker here 👍

Thanks for the information. Could you please share details behind this decision? I'm wondering if it's because of the fact that TF service deployer is not feasible for your use cases.

Another thought: do you plan to move `akroh/stream` under `elastic` org? I know it's Andrew's repository, but I believe in general we should depend on company's ownership.

**Comment by andrewkroh** [comment]
> Thanks for the information. Could you please share details behind this decision? I'm wondering if it's because of the fact that TF service deployer is not feasible for your use cases.

I think it would be valuable to have a test like this too. It would serve as an integration test against the real service and could catch different types of issues than the emulator based testing with static logs. If the TF setup configures the actual services to write logs that's even better since it would let us catch when log formats changes. And lastly it could service as a pseudo reference for describing the configuration steps necessary to setup GCP to use the package.

> Another thought: do you plan to move akroh/stream under elastic org? I know it's Andrew's repository, but I believe in general we should depend on company's ownership

I agree. I wasn't sure whether it was going to be useful at first so I opted for a personal repo where I could more independently setup CI. I'll work on get this moved over and have Jenkins setup.

File: packages/gcp/data_stream/audit/manifest.yml [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        multi: false
+        required: true
+        show_user: true
+        default: false
+      - name: tags
+        type: text
+        title: Tags
+        multi: true
+        required: true
+        show_user: true
```
Comment (andrewkroh) [inline packages/gcp/data_stream/audit/manifest.yml:34]: Can you set tags to `show_user: false` across them all. I prefer to have some of the non-essential/advanced settings not displayed by default.

File: go.sum [hunk 3]
```diff
@@ -78,7 +78,6 @@ github.com/cpuguy83/go-md2man/v2 v2.0.0/go.mod h1:maD7wRr/U5Z6m/iR4s+kqSMx2CaBsr
 github.com/creack/pty v1.1.9/go.mod h1:oKZEueFk5CKHvIhNR5MUki03XCEU+Q6VDXinZuGJ33E=
 github.com/creasty/defaults v1.5.1 h1:j8WexcS3d/t4ZmllX4GEkl4wIB/trOr035ajcLHCISM=
 github.com/creasty/defaults v1.5.1/go.mod h1:FPZ+Y0WNrbqOVw+c6av63eyHUAl6pMHZwqLPvXUZGfY=
-github.com/davecgh/go-spew v1.1.0 h1:ZDRjVQ15GmhC3fiQ8ni8+OwkZQO4DARzQgrnXU1Liz8=
```
Comment (andrewkroh) [inline go.sum:81]: Is this changed needed?

**Review by andrewkroh — APPROVED** [review]
LGTM. Nice work with all the testing improvements to allow the emulator to be used.

**Pull request merged**